lib/messageReceiver.ts

```diff
@@ -0,0 +1,232 @@
+   * @property {Receiver} [_receiver] The AMQP receiver link.
+   * @protected
+   */
+  protected _receiver?: Receiver;
+  /**
+   * @property {OnMessage} _onMessage The message handler provided by the user that will be wrapped
+   * inside _onAmqpMessage.
+   * @protected
+   */
+  protected _onMessage?: OnMessage;
+  /**
+   * @property {OnMessage} _onMessage The error handler provided by the user that will be wrapped
+   * inside _onAmqpError.
+   * @protected
+   */
+  protected _onError?: OnError;
+  /**
+   * @property {OnMessage} _onMessage The message handler that will be set as the handler on the
+   * underlying rhea receiver for the "message" event.
+   * @protected
+   */
+  protected _onAmqpMessage: OnAmqpEvent;
+  /**
+   * @property {OnMessage} _onMessage The message handler that will be set as the handler on the
+   * underlying rhea receiver for the "receiver_error" event.
+   * @protected
+   */
+  protected _onAmqpError: OnAmqpEvent;
+
+  constructor(context: ClientEntityContext, receiverType: ReceiverType, options?: ReceiveOptions) {
+    super(`${context.entityPath}`, context);
+    if (!options) options = {};
+    this.receiverType = receiverType;
+    this.address = `${this._context.entityPath}`;
+    this.audience = `${this._context.namespace.config.endpoint}${this._context.entityPath}`;
+    this.maxConcurrentCalls = options.maxConcurrentCalls != undefined ?
+      options.maxConcurrentCalls : 1;
+    this.autoComplete = options.autoComplete != undefined ? options.autoComplete : false;
+    this.receiveMode = options.receiveMode || ReceiveMode.peekLock;
+    this._onAmqpMessage = async (context: EventContext) => {
+      const bMessage = new Message(context.message!, context.delivery!);
+      bMessage.body = this._context.namespace.dataTransformer.decode(context.message!.body);
+      try {
+        await this._onMessage!(bMessage);
+        if (this.autoComplete) {
+          debug("[%s] Auto completing the message with id '%s' on the receiver '%s'.",
+            this._context.namespace.connectionId, bMessage.messageId, this.id);
+          bMessage.complete();
+        }
+      } catch (err) {
+        debug("[%s] Abandoning the message with id '%s' on the receiver '%s' since an error " +
+          "occured: %O.",
+          this._context.namespace.connectionId, bMessage.messageId, this.id);
+        bMessage.abandon();
+      }
+    };
+
+    this._onAmqpError = (context: EventContext) => {
+      const sbError = translate(context.receiver!.error!);
+      // TODO: Should we retry before calling user's error method?
+      debug("[%s] An error occurred for Receiver '%s': %O.",
+        this._context.namespace.connectionId, this.id, sbError);
+      this._onError!(sbError);
+    };
+  }
+
+  /**
+   * Determines whether the AMQP receiver link is open. If open then returns true else returns false.
+   * @returns {boolean} boolean
+   */
+  isOpen(): boolean {
+    return this._receiver! && this._receiver!.isOpen();
```

Comment:
nit: the `!` undermines the check you're making from a type perspective. just do `this._receiver && this._receiver.isOpen()` the first check is enough for typescript to infer that `this._receiver` exists for the second one

Reply:
Without the excalamation mark, I see the following error
> Type 'boolean | undefined' is not assignable to type 'boolean'.
  Type 'undefined' is not assignable to type 'boolean'.

Transforming to this `return this._receiver ? this._receiver.isOpen() : false;` makes tsc happy.